embedded-graphics/src/primitives/ellipse.rs

```diff
@@ -0,0 +1,459 @@
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    iter: Points,
+    outer_size: Size,
+    outer_color: Option<C>,
+    inner_size: Size,
+    inner_color: Option<C>,
+    center: Point,
+    inner_width: u32,
+}
+
+impl<C> StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    fn new(styled: &Styled<Ellipse, PrimitiveStyle<C>>) -> Self {
+        // Always use a stroke width of 0 if no stroke color was set.
+        let stroke_width = styled.style.effective_stroke_width();
+
+        let outer_size = styled.primitive.size;
+        let inner_size = outer_size.saturating_sub(Size::new(2 * stroke_width, 2 * stroke_width));
+
+        let inner_width = inner_size.width;
+
+        let inner_size = Size::new(inner_size.width.pow(2), inner_size.height.pow(2));
+        let iter = if !styled.style.is_transparent() {
+            Points::new(&styled.primitive)
+        } else {
+            Points::empty()
+        };
+
+        Self {
+            iter,
+            outer_size,
+            outer_color: styled.style.stroke_color,
+            inner_size,
+            inner_color: styled.style.fill_color,
+            center: styled.primitive.center_2x(),
+            inner_width,
+        }
+    }
+}
+
+impl<C> Iterator for StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    type Item = Pixel<C>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let Self {
+            inner_size,
+            inner_color,
+            outer_color,
+            center,
+            inner_width,
+            ..
+        } = *self;
+
+        self.iter.find_map(|point| {
+            let inner_threshold = size_to_threshold(inner_size, point * 2 - center, inner_width);
+
+            let color = if inner_threshold {
+                inner_color
+            } else {
+                outer_color
+            };
+
+            color.map(|color| Pixel(point, color))
+        })
+    }
+}
+
+impl<'a, C: 'a> Drawable<C> for &Styled<Ellipse, PrimitiveStyle<C>>
+where
+    C: PixelColor,
+{
+    fn draw<D: DrawTarget<C>>(self, display: &mut D) -> Result<(), D::Error> {
+        display.draw_iter(self)
+    }
+}
+
+/// Uses the ellipse equation b^2 * x^2 + a^2 * y^2 - a^2 * b^2 to return a value signifying whether
+/// a given point lies inside (`true`) or outside (`false`) an ellipse centered around `(0, 0)` with
+/// width and height defined by the `size` parameter.
+fn size_to_threshold(size: Size, point: Point, diameter: u32) -> bool {
+    let Size {
+        width: a,
+        height: b,
+    } = size;
+
+    let Point { x, y } = point;
+
+    let x = x.pow(2);
+    let y = y.pow(2);
+
+    // Special case for circles, where width and height are equal
+    if a == b {
+        let threshold = if diameter <= 4 {
+            diameter.pow(2) - diameter / 2
+        } else {
+            diameter.pow(2)
+        } as i32;
```

Comment:
You can refactor this code by computing this in the constructor and pass `threshold` instead of `diameter`. For the non circle case you can rearrange the formula to ` b * x + a * y <  threshold` and pass `a * b` as `threshold`.